cli/scorecard/violations.go

```diff
@@ -135,6 +138,74 @@ func getViolations(inventory *InventoryConfig, config *ScoringConfig) ([]*RichVi
 	return richViolations, nil
 }
 
+// getViolationsConcurrently finds all Config Validator violations for a given Inventory concurrently
+func getViolationsConcurrently(inventory *InventoryConfig, config *ScoringConfig) ([]*RichViolation, error) {
+	var err error
+	var pbAssets []*validator.Asset
+
+	if inventory.bucketName != "" {
+		pbAssets, err = getDataFromBucket(inventory.bucketName)
+		if err != nil {
+			return nil, errors.Wrap(err, "Fetching inventory from Bucket")
+		}
+	} else if inventory.dirPath != "" {
+		pbAssets, err = getDataFromFile(inventory.dirPath)
+		if err != nil {
+			return nil, errors.Wrap(err, "Fetching inventory from local directory")
+		}
+	} else if inventory.readFromStdin {
+		pbAssets, err = getDataFromStdin()
+		if err != nil {
+			return nil, errors.Wrap(err, "Reading from stdin")
+		}
+	}
+
+	richViolations := make([]*RichViolation, 0)
+	wp := workerpool.New(numWorkers())
+	var badAsset *validator.Asset
+	for _, asset := range pbAssets {
+		asset := asset
+		wp.Submit(func() {
+			violations, errAsset := config.validator.ReviewAsset(context.Background(), asset)
+			if errAsset != nil {
+				err = errAsset
+				badAsset = asset
+				wp.Stop()
+			}
+			for _, violation := range violations {
+				richViolation := RichViolation{*violation, "", violation.Resource, violation.Message, violation.Metadata, asset}
```

Comment:
This doesn't look thread-safe as different workers will attempt to add to the same array. You can run go with `-race` to verify. Results likely need to be returned via channel.

Reply:
I have added mutex lock and unlock to append

Reply:
-race does not show any warnings now after mutex addition

Reply:
This should work, but I'm not sure a mutex is best here. I still think it might be better to have results via a channel so no locking is needed.

Reply:
I have tried the channel approach before, see below: its not giving the required performance, perhaps maybe because this is my first time with golang. I might be doing it wrongly. With mutext I havent seen any performance degradation

```
	richViolations := make([]*RichViolation, 0)
	wp := workerpool.New(inventory.workers)
	var badAsset *validator.Asset
	for _, asset := range pbAssets {
		asset := asset
		richViolationChannel := make(chan *RichViolation)
		wp.Submit(func() {
			violations, errAsset := config.validator.ReviewAsset(context.Background(), asset)
			if errAsset != nil {
				err = errAsset
				badAsset = asset
				wp.Stop()
			}
			for _, violation := range violations {
				richViolation := RichViolation{*violation, "", violation.Resource, violation.Message, violation.Metadata, asset}
				richViolationChannel <- &richViolation
				//richViolations = append(richViolations, &richViolation)
			}
			close(richViolationChannel)
		})
		//close(richViolationChannel)
		for violation := range richViolationChannel {
			richViolations = append(richViolations, violation)
		}
		richViolationChannel = make(chan *RichViolation)
	}
	wp.StopWait()
```

Reply:
I think the problem is you are creating the channel and reading from it within the pbAssets for loop. Can you try something like this?

```
richViolations := make([]*RichViolation, 0)
wp := workerpool.New(inventory.workers)
richViolationChannel := make(chan *RichViolation, len(pbAssets))

var badAsset *validator.Asset

for _, asset := range pbAssets {
  wp.Submit(func() {
    violations, errAsset := config.validator.ReviewAsset(context.Background(), asset)
    if errAsset != nil {
      err = errAsset
      badAsset = asset
      wp.Stop()
    }
    for _, violation := range violations {
      richViolation := RichViolation{*violation, "", violation.Resource, violation.Message, violation.Metadata, asset}
      richViolationChannel <- &richViolation
    }
  })
}

go func() {
  for violation := range richViolationChannel {
    richViolations = append(richViolations, violation)
  }
}

wp.StopWait()
close(richViolationChannel)
```

Reply:
yea  I had tried this approach before and again now, it just loops over quickly and quits with the following message:

Generating CFT scorecard
No issues found found! You have a perfect score.

Reply:
The problem could be the go func(){}() around the append loop?

Reply:
so when I run that code I also get race conditions:
```
 $ time go run --race main.go scorecard --policy-path /Users/hak/Desktop/Development/GCP/gcp-policy-library/policy-library --dir-path /Users/hak/Desktop/Development/GCP/CFT-master/cloud-foundation-toolkit/cli/testdata/scorecard/cai-dir --output-path . --output-format csv --workers 4
Generating CFT scorecard
==================
WARNING: DATA RACE
Read at 0x00c000ae2058 by goroutine 39:
  github.com/GoogleCloudPlatform/cloud-foundation-toolkit/cli/scorecard.getViolations.func1()
      /Users/hak/Desktop/Development/GCP/myfork/cloud-foundation-toolkit/cli/scorecard/violations.go:133 +0xe9
  github.com/gammazero/workerpool.startWorker()
      /Users/hak/go/pkg/mod/github.com/gammazero/workerpool@v1.1.2/workerpool.go:233 +0x34

Previous write at 0x00c000ae2058 by main goroutine:
  github.com/GoogleCloudPlatform/cloud-foundation-toolkit/cli/scorecard.getViolations()
      /Users/hak/Desktop/Development/GCP/myfork/cloud-foundation-toolkit/cli/scorecard/violations.go:131 +0x33e
  github.com/GoogleCloudPlatform/cloud-foundation-toolkit/cli/scorecard.(*InventoryConfig).findViolations()
      /Users/hak/Desktop/Development/GCP/myfork/cloud-foundation-toolkit/cli/scorecard/score.go:261 +0x46
  github.com/GoogleCloudPlatform/cloud-foundation-toolkit/cli/scorecard.(*InventoryConfig).Score()
      /Users/hak/Desktop/Development/GCP/myfork/cloud-foundation-toolkit/cli/scorecard/score.go:275 +0x64
  github.com/GoogleCloudPlatform/cloud-foundation-toolkit/cli/scorecard.glob..func2()
      /Users/hak/Desktop/Development/GCP/myfork/cloud-foundation-toolkit/cli/scorecard/cmd.go:110 +0x586
  github.com/spf13/cobra.(*Command).execute()
      /Users/hak/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:826 +0x505
  github.com/spf13/cobra.(*Command).ExecuteC()
      /Users/hak/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914 +0x431
  github.com/spf13/cobra.(*Command).Execute()
      /Users/hak/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864 +0x4a
  github.com/GoogleCloudPlatform/cloud-foundation-toolkit/cli/cmd.Execute()
      /Users/hak/Desktop/Development/GCP/myfork/cloud-foundation-toolkit/cli/cmd/root.go:75 +0x2b
  main.main()
      /Users/hak/Desktop/Development/GCP/myfork/cloud-foundation-toolkit/cli/main.go:7 +0x2f

Goroutine 39 (running) created at:
  github.com/gammazero/workerpool.(*WorkerPool).dispatch()
      /Users/hak/go/pkg/mod/github.com/gammazero/workerpool@v1.1.2/workerpool.go:195 +0x470
==================
==================
WARNING: DATA RACE
Read at 0x00c00053d880 by goroutine 41:
  github.com/forseti-security/config-validator/pkg/api/validator.(*Asset).GetAncestryPath()
      /Users/hak/go/pkg/mod/github.com/forseti-security/config-validator@v0.0.0-20200505040130-17dc60b21dc8/pkg/api/validator/validator.pb.go:110 +0x551
  github.com/forseti-security/config-validator/pkg/asset.ValidateAsset()
      /Users/hak/go/pkg/mod/github.com/forseti-security/config-validator@v0.0.0-20200505040130-17dc60b21dc8/pkg/asset/asset.go:43 +0x80
  github.com/forseti-security/config-validator/pkg/gcv.(*Validator).ReviewAsset()
      /Users/hak/go/pkg/mod/github.com/forseti-security/config-validator@v0.0.0-20200505040130-17dc60b21dc8/pkg/gcv/validator.go:155 +0x55
  github.com/GoogleCloudPlatform/cloud-foundation-toolkit/cli/scorecard.getViolations.func1()
      /Users/hak/Desktop/Development/GCP/myfork/cloud-foundation-toolkit/cli/scorecard/violations.go:133 +0x124
  github.com/gammazero/workerpool.startWorker()
      /Users/hak/go/pkg/mod/github.com/gammazero/workerpool@v1.1.2/workerpool.go:233 +0x34

Previous write at 0x00c00053d880 by goroutine 40:
  github.com/forseti-security/config-validator/pkg/asset.SanitizeAncestryPath()
      /Users/hak/go/pkg/mod/github.com/forseti-security/config-validator@v0.0.0-20200505040130-17dc60b21dc8/pkg/asset/asset.go:82 +0xc8
  github.com/forseti-security/config-validator/pkg/gcv.(*Validator).ReviewAsset()
      /Users/hak/go/pkg/mod/github.com/forseti-security/config-validator@v0.0.0-20200505040130-17dc60b21dc8/pkg/gcv/validator.go:159 +0x7c
  github.com/GoogleCloudPlatform/cloud-foundation-toolkit/cli/scorecard.getViolations.func1()
      /Users/hak/Desktop/Development/GCP/myfork/cloud-foundation-toolkit/cli/scorecard/violations.go:133 +0x124
  github.com/gammazero/workerpool.startWorker()
      /Users/hak/go/pkg/mod/github.com/gammazero/workerpool@v1.1.2/workerpool.go:233 +0x34

Goroutine 41 (running) created at:
  github.com/gammazero/workerpool.(*WorkerPool).dispatch()
      /Users/hak/go/pkg/mod/github.com/gammazero/workerpool@v1.1.2/workerpool.go:195 +0x470

Goroutine 40 (running) created at:
  github.com/gammazero/workerpool.(*WorkerPool).dispatch()
      /Users/hak/go/pkg/mod/github.com/gammazero/workerpool@v1.1.2/workerpool.go:195 +0x470
==================
No issues found found! You have a perfect score.
Found 2 data race(s)
exit status 66
go run --race main.go scorecard --policy-path  --dir-path  --output-path .     13.22s user 2.47s system 125% cpu 12.534 total
```

Reply:
Interesting... if you comment out the `for violation := range richViolationChannel {` block does `-race` succeed? I haven't used the worker pool package before so it might not work well with returning results via a channel.

If you prefer to stick with the mutex approach, I'm fine to do that as well. It's not quite idiomatic but should be decently performant.

Reply:
I'd stick with the Mutex for now until we find a better approach. If I comment that section out I still get the race condition.

```
 time go run --race main.go scorecard --policy-path /Users/hak/Desktop/Development/GCP/gcp-policy-library/policy-library --dir-path /Users/hak/Desktop/Development/GCP/CFT-master/cloud-foundation-toolkit/cli/testdata/scorecard/cai-dir --output-path . --output-format csv --workers 4
Generating CFT scorecard
==================
WARNING: DATA RACE
Read at 0x00c0007d0798 by goroutine 39:
  github.com/GoogleCloudPlatform/cloud-foundation-toolkit/cli/scorecard.getViolations.func1()
      /Users/hak/Desktop/Development/GCP/myfork/cloud-foundation-toolkit/cli/scorecard/violations.go:133 +0xe9
  github.com/gammazero/workerpool.startWorker()
      /Users/hak/go/pkg/mod/github.com/gammazero/workerpool@v1.1.2/workerpool.go:233 +0x34

Previous write at 0x00c0007d0798 by main goroutine:
  github.com/GoogleCloudPlatform/cloud-foundation-toolkit/cli/scorecard.getViolations()
      /Users/hak/Desktop/Development/GCP/myfork/cloud-foundation-toolkit/cli/scorecard/violations.go:131 +0x2ee
  github.com/GoogleCloudPlatform/cloud-foundation-toolkit/cli/scorecard.(*InventoryConfig).findViolations()
      /Users/hak/Desktop/Development/GCP/myfork/cloud-foundation-toolkit/cli/scorecard/score.go:261 +0x46
  github.com/GoogleCloudPlatform/cloud-foundation-toolkit/cli/scorecard.(*InventoryConfig).Score()
      /Users/hak/Desktop/Development/GCP/myfork/cloud-foundation-toolkit/cli/scorecard/score.go:275 +0x64
  github.com/GoogleCloudPlatform/cloud-foundation-toolkit/cli/scorecard.glob..func2()
      /Users/hak/Desktop/Development/GCP/myfork/cloud-foundation-toolkit/cli/scorecard/cmd.go:110 +0x586
  github.com/spf13/cobra.(*Command).execute()
      /Users/hak/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:826 +0x505
  github.com/spf13/cobra.(*Command).ExecuteC()
      /Users/hak/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914 +0x431
  github.com/spf13/cobra.(*Command).Execute()
      /Users/hak/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864 +0x4a
  github.com/GoogleCloudPlatform/cloud-foundation-toolkit/cli/cmd.Execute()
      /Users/hak/Desktop/Development/GCP/myfork/cloud-foundation-toolkit/cli/cmd/root.go:75 +0x2b
  main.main()
      /Users/hak/Desktop/Development/GCP/myfork/cloud-foundation-toolkit/cli/main.go:7 +0x2f

Goroutine 39 (running) created at:
  github.com/gammazero/workerpool.(*WorkerPool).dispatch()
      /Users/hak/go/pkg/mod/github.com/gammazero/workerpool@v1.1.2/workerpool.go:195 +0x470
==================
==================
WARNING: DATA RACE
Read at 0x00c00051f240 by goroutine 39:
  github.com/forseti-security/config-validator/pkg/api/validator.(*Asset).GetAncestryPath()
      /Users/hak/go/pkg/mod/github.com/forseti-security/config-validator@v0.0.0-20200505040130-17dc60b21dc8/pkg/api/validator/validator.pb.go:110 +0x551
  github.com/forseti-security/config-validator/pkg/asset.ValidateAsset()
      /Users/hak/go/pkg/mod/github.com/forseti-security/config-validator@v0.0.0-20200505040130-17dc60b21dc8/pkg/asset/asset.go:43 +0x80
  github.com/forseti-security/config-validator/pkg/gcv.(*Validator).ReviewAsset()
      /Users/hak/go/pkg/mod/github.com/forseti-security/config-validator@v0.0.0-20200505040130-17dc60b21dc8/pkg/gcv/validator.go:155 +0x55
  github.com/GoogleCloudPlatform/cloud-foundation-toolkit/cli/scorecard.getViolations.func1()
      /Users/hak/Desktop/Development/GCP/myfork/cloud-foundation-toolkit/cli/scorecard/violations.go:133 +0x124
  github.com/gammazero/workerpool.startWorker()
      /Users/hak/go/pkg/mod/github.com/gammazero/workerpool@v1.1.2/workerpool.go:233 +0x34

Previous write at 0x00c00051f240 by goroutine 41:
  github.com/forseti-security/config-validator/pkg/asset.SanitizeAncestryPath()
      /Users/hak/go/pkg/mod/github.com/forseti-security/config-validator@v0.0.0-20200505040130-17dc60b21dc8/pkg/asset/asset.go:82 +0xc8
  github.com/forseti-security/config-validator/pkg/gcv.(*Validator).ReviewAsset()
      /Users/hak/go/pkg/mod/github.com/forseti-security/config-validator@v0.0.0-20200505040130-17dc60b21dc8/pkg/gcv/validator.go:159 +0x7c
  github.com/GoogleCloudPlatform/cloud-foundation-toolkit/cli/scorecard.getViolations.func1()
      /Users/hak/Desktop/Development/GCP/myfork/cloud-foundation-toolkit/cli/scorecard/violations.go:133 +0x124
  github.com/gammazero/workerpool.startWorker()
      /Users/hak/go/pkg/mod/github.com/gammazero/workerpool@v1.1.2/workerpool.go:233 +0x34

Goroutine 39 (running) created at:
  github.com/gammazero/workerpool.(*WorkerPool).dispatch()
      /Users/hak/go/pkg/mod/github.com/gammazero/workerpool@v1.1.2/workerpool.go:195 +0x470

Goroutine 41 (running) created at:
  github.com/gammazero/workerpool.(*WorkerPool).dispatch()
      /Users/hak/go/pkg/mod/github.com/gammazero/workerpool@v1.1.2/workerpool.go:195 +0x470
==================
No issues found found! You have a perfect score.
Found 2 data race(s)
exit status 66
go run --race main.go scorecard --policy-path  --dir-path  --output-path .     13.66s user 2.68s system 123% cpu 13.238 total
```

Reply:
Let me know your thoughts.. otherwise lets considered this resolved and merge?